File: source/downloads/code/LeetCode/Python/Combination-Sum.py
class Solution(object):
    def recur_helper(self,c,target,result,vector,cur):
        if target == 0:
            result.append(vector[:])
            return
        
        while cur<len(c):
            if c[cur] <= target:
                vector.append(c[cur])
                self.recur_helper(c,target-c[cur],result,vector,cur)
                vector.pop()
            cur += 1
        return

    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        
        result = []
        candidates = sorted(candidates)
        if candidates == None or len(candidates) == 0:
            return result
        self.recur_helper(candidates,target,result,[],0)
        return result

File: source/downloads/code/LeetCode/Java/Combination-Sum.java
public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        List<Integer> tmp = new ArrayList<Integer>();
        Arrays.sort(candidates);
        dfs(ret, tmp, candidates, target, 0);
        return ret;        
    }

    void dfs(List<List<Integer>> ret, List<Integer> tmp, int[] candidates, int target, int start){
        if(target==0)
            ret.add(new ArrayList<Integer>(tmp));
        else if (target>0){
            for(int i= start; i < candidates.length; i++){
                tmp.add(candidates[i]);
                dfs(ret, tmp, candidates, target-candidates[i], i);
                tmp.remove(tmp.size()-1);
            }
        }
    }
}
